prow/cmd/branchprotector/protect.go

```diff
@@ -267,18 +273,97 @@ func (p *protector) UpdateRepo(orgName string, repoName string, repo config.Repo
 		}
 	}
 
+	var collaborators, teams []string
+	if p.verifyAuthorisation {
+		collaborators, err = p.authorizedCollaborators(orgName, repoName)
+		if err != nil {
+			logrus.Infof("%s/%s: error getting list of collaborators: %v", orgName, repoName, err)
+			return err
+		}
+
+		teams, err = p.authorizedTeams(orgName, repoName)
+		if err != nil {
+			logrus.Infof("%s/%s: error getting list of teams: %v", orgName, repoName, err)
+			return err
+		}
+	}
+
 	for bn, githubBranch := range branches {
 		if branch, err := repo.GetBranch(bn); err != nil {
 			return fmt.Errorf("get %s: %v", bn, err)
-		} else if err = p.UpdateBranch(orgName, repoName, bn, *branch, githubBranch.Protected); err != nil {
+		} else if err = p.UpdateBranch(orgName, repoName, bn, *branch, githubBranch.Protected, collaborators, teams); err != nil {
 			return fmt.Errorf("update %s from protected=%t: %v", bn, githubBranch.Protected, err)
 		}
 	}
 	return nil
 }
 
+// authorizedCollaborators returns the list of Logins for users that are
+// authorized to write to a repository.
+func (p *protector) authorizedCollaborators(org, repo string) ([]string, error) {
+	collaborators, err := p.client.ListCollaborators(org, repo)
+	if err != nil {
+		return nil, err
+	}
+	var authorized []string
+	for _, c := range collaborators {
+		if c.Permissions.Admin || c.Permissions.Push {
+			authorized = append(authorized, c.Login)
+		}
+	}
+	return authorized, nil
+}
+
+// authorizedTeams returns the list of slugs for teams that are authorized to
+// write to a repository.
+func (p *protector) authorizedTeams(org, repo string) ([]string, error) {
+	teams, err := p.client.ListRepoTeams(org, repo)
+	if err != nil {
+		return nil, err
+	}
+	var authorized []string
+	for _, t := range teams {
+		if t.Permission == github.RepoPush || t.Permission == github.RepoAdmin {
+			authorized = append(authorized, t.Slug)
+		}
+	}
+	return authorized, nil
+}
+
+func validateRequest(org, repo string, bp *github.BranchProtectionRequest, authorizedCollaborators, authorizedTeams []string) []error {
```

Comment:
We're not checking if a user has write access by virtue of being in a team that has it here

Reply:
Thanks for the review. From [the `ListCollaborators` section of the github API page](https://developer.github.com/v3/repos/collaborators/#list-collaborators) it seems the `ListCollaborators` call that we make accounts for this. (we also use the `application/vnd.github.hellcat-preview+json` header, thus including members of nested teams as well).